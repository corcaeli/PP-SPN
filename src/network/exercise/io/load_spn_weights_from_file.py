from network.exercise.exercise_class import Exercise
from globals import IDs, Keys


def add_exercise_load_spn_weights_from_file(manager):
    manager.exercises.append(Exercise(IDs.LOAD_SPN_WEIGHTS_FROM_FILE))


def load_spn_weights_from_file(member, message_value):
    spn_weights_file_path = member.config[f"ID_{member.id}"].get(
        Keys.CONFIG_SPN_WEIGHTS_FILE_PATH
    )
    f = open(spn_weights_file_path, "r")
    for line in f.read().splitlines():
        data_id, value_str = line.split("=")
        value = int(value_str)

        member.data[data_id] = value
    f.close()
    member.network_socket.send(
        member.manager_id_chip, IDs.LOAD_SPN_WEIGHTS_FROM_FILE, member.id
    )
