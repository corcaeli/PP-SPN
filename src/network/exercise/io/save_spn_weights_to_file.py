from spn.structure.Base import Sum, get_nodes_by_type
from network.exercise.exercise_class import Exercise
from globals import IDs, Keys


def add_exercise_save_spn_weights_to_file(manager):
    manager.exercises.append(Exercise(IDs.SAVE_SPN_WEIGHTS_TO_FILE))


def save_spn_weights_to_file(member, message_value):
    spn_weights_file_path = member.config[f"ID_{member.id}"].get(
        Keys.CONFIG_SPN_WEIGHTS_FILE_PATH
    )
    f = open(spn_weights_file_path, "w")
    for sum_node in get_nodes_by_type(member.spn, Sum):
        for child_node in sum_node.children:
            data_id_weight = f"({sum_node.id}, {child_node.id})"
            value = member.data.get(data_id_weight)
            f.write(f"{data_id_weight}={value}\n")

    f.close()
    member.network_socket.send(
        member.manager_id_chip, IDs.SAVE_SPN_WEIGHTS_TO_FILE, member.id
    )
