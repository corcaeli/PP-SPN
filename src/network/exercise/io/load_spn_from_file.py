from network.exercise.io.save_spn_weights_to_file import (
    add_exercise_save_spn_weights_to_file,
)
from spn.algorithms.layerwise.layers import Sum
from network.exercise.exercise_class import Exercise
from globals import IDs, Keys

from spn.structure.Base import get_nodes_by_type, get_topological_order_layers

from functools import partial


def add_exercise_load_spn_from_file(manager):
    manager.exercises.append(
        Exercise(
            IDs.LOAD_SPN_FROM_FILE,
            on_completion=partial(on_completion_load_spn_from_file, manager),
        )
    )


def on_completion_load_spn_from_file(manager, exercise):
    manager.default_on_completion(exercise)
    # from spn.io.Graphics import plot_spn
    # plot_spn(manager.spn, "plots/paper_spn_2.png")
    # add_exercise_save_spn_weights_to_file(manager)
    from spn.algorithms.Statistics import get_structure_stats

    print(get_structure_stats(manager.spn))


def load_spn_from_file(member, message_value=None):
    spn_file_path = member.config[f"ID_{member.id}"].get(Keys.CONFIG_SPN_FILE_PATH)
    assert (
        spn_file_path is not None
    ), f'No "{Keys.CONFIG_SPN_FILE_PATH}" given in the config file'
    f = open(spn_file_path, "r")
    spn_str = f.read()
    member.spn = member.str_to_spn(spn_str)

    member.layers = get_topological_order_layers(member.spn)

    member.network_socket.send(
        member.manager_id_chip, IDs.LOAD_SPN_FROM_FILE, member.id
    )
