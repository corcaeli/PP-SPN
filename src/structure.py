import numpy as np
from spn.structure.Base import Node

from example_spn import example_spn_from_paper_about_spn

# TODO: to implement
def build_spn_structure(data) -> Node:
    assert data is not None
    return example_spn_from_paper_about_spn()


from spn.structure.Base import get_nodes_by_type
from spn.structure.Base import Sum


def change_stuf(node):
    node.weights[0] = 1


def replace_sum_node_weights(spn_root_node, replacement_value=1) -> None:
    sum_nodes = get_nodes_by_type(spn_root_node, Sum)
    for sum_node in sum_nodes:
        sum_node.weights = np.full((len(sum_node.weights)), replacement_value)
        # for index in len(sum_node.weights):
        #    sum_node.weights[index] = 1
    return spn_root_node


def stuf():
    np.random.seed(123)
    train_data = np.c_[
        np.r_[np.random.normal(5, 1, (500, 2)), np.random.normal(10, 1, (500, 2))],
        np.r_[np.zeros((500, 1)), np.ones((500, 1))],
    ]

    from spn.algorithms.LearningWrappers import learn_parametric, learn_classifier
    from spn.structure.leaves.parametric.Parametric import Categorical, Gaussian
    from spn.structure.Base import Context

    spn_classification = learn_classifier(
        train_data,
        Context(parametric_types=[Gaussian, Gaussian, Categorical]).add_domains(
            train_data
        ),
        learn_parametric,
        2,
    )
    # print(train_data)
    test_classification = np.array([3.0, 4.0, np.nan, 12.0, 18.0, np.nan]).reshape(
        -1, 3
    )

    from spn.algorithms.MPE import mpe

    print(mpe(spn_classification, test_classification))

    a = np.random.randint(2, size=1000).reshape(-1, 1)
    b = np.random.randint(3, size=1000).reshape(-1, 1)
    c = np.r_[np.random.normal(10, 5, (300, 1)), np.random.normal(20, 10, (700, 1))]
    d = 5 * a + 3 * b + c
    train_data = np.c_[a, b, c, d]
    # print(d)
    print("=========================")
    print(train_data)

    # MSPN
    from spn.structure.Base import Context
    from spn.structure.StatisticalTypes import MetaType

    # ds_context = Context(meta_types=[MetaType.DISCRETE, MetaType.DISCRETE, MetaType.REAL, MetaType.REAL])
    # ds_context.add_domains(train_data)

    from spn.algorithms.LearningWrappers import learn_mspn

    # mspn = learn_mspn(train_data, ds_context, min_instances_slice=20)

    from spn.io.Graphics import plot_spn

    # plot_spn(mspn, 'plots/mspn.png')

    # parametric SPN
    from spn.structure.Base import Context
    from spn.structure.leaves.parametric.Parametric import Categorical, Gaussian

    ds_context = Context(
        parametric_types=[Categorical, Categorical, Gaussian, Gaussian]
    ).add_domains(train_data)
    spn_classification = learn_classifier(
        train_data,
        Context(
            parametric_types=[Categorical, Categorical, Gaussian, Gaussian]
        ).add_domains(train_data),
        learn_parametric,
        2,
    )

    from spn.algorithms.LearningWrappers import learn_parametric

    # spn = learn_parametric(train_data, ds_context, min_instances_slice=20)
    # plot_spn(spn, 'plots/spn.png')
