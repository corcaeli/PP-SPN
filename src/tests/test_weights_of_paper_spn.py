from spn.algorithms.Inference import likelihood
from spn.structure.leaves.parametric.Parametric import Categorical

from spn.structure.Base import Sum, Product, eval_spn_bottom_up, get_nodes_by_type
from spn.structure.Base import assign_ids, rebuild_scopes_bottom_up


def example_spn_from_paper_about_spn():
    # scope 0 = a
    a = Categorical(p=[0.0, 1.0], scope=0)
    a_not = Categorical(p=[1.0, 0.0], scope=0)

    # scope 1 = b
    b = Categorical(p=[0.0, 1.0], scope=1)
    b_not = Categorical(p=[1.0, 0.0], scope=1)

    # scope 2 = c
    c = Categorical(p=[0.0, 1.0], scope=2)
    c_not = Categorical(p=[1.0, 0.0], scope=2)

    n_14 = Sum(weights=[0.1, 0.9], children=[c, c_not])  # 0.1, 0.9
    n_15 = Sum(weights=[0.8, 0.2], children=[c, c_not])
    n_16 = Sum(weights=[0.3, 0.7], children=[c, c_not])

    n_8 = Product(children=[n_14, b])
    n_9 = Product(children=[n_15, b_not])
    n_10 = Product(children=[b, n_16])
    n_11 = Product(children=[b_not, n_16])

    n_6 = Sum(weights=[0.4, 0.6], children=[n_8, n_9])
    n_7 = Sum(weights=[0.5, 0.5], children=[n_10, n_11])

    n_2 = Product(children=[a, n_6])
    n_3 = Product(children=[n_7, a_not])

    n_1 = Sum(weights=[0.3, 0.7], children=[n_2, n_3])

    spn = n_1

    assign_ids(spn)
    rebuild_scopes_bottom_up(spn)

    from spn.io.Text import spn_to_str_equation
    from spn.io.Text import str_to_spn

    spn_str = spn_to_str_equation(spn)
    spn = str_to_spn(spn_str)

    return spn


import numpy as np

spn = example_spn_from_paper_about_spn()
input = np.array([[0, 1, 0]])
result = likelihood(spn, input)
print(result)  # 0.245

input = np.array([[1, 1, 0]])
result = likelihood(spn, input)
print(result)  # 0.108
# likelihood(spn, [input])
from spn.io.Graphics import plot_spn

plot_spn(spn, "spn_rebuild.png")


# f = open("./tests/test_weights_of_paper_spn.out", "w")
# data = {}
# for sum_node in get_nodes_by_type(spn, Sum):
#    children = sum_node.children
#    for index in range(len(children)):
#        child_node = children[index]
#        data_id_weight = f"({sum_node.id}, {child_node.id})"
#        # value = data.get(data_id_weight)
#        value = sum_node.weights[index]
#        f.write(f"{data_id_weight}={value}\n")
