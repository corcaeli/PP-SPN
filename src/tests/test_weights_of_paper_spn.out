(0, 1)=0.3
(0, 2)=0.7
(4, 7)=0.4
(4, 8)=0.6
(5, 9)=0.5
(5, 10)=0.5
(11, 19)=0.1
(11, 20)=0.9
(13, 21)=0.8
(13, 22)=0.2
(16, 23)=0.3
(16, 24)=0.7
(18, 25)=0.3
(18, 26)=0.7
